lib/rex/socket.rb

```diff
@@ -809,10 +828,64 @@ def type?
 
 protected
 
-  attr_writer :peerhost, :peerport, :localhost, :localport # :nodoc:
+  attr_writer :peerhost, :peerhostname, :peerport, :localhost, :localport # :nodoc:
   attr_writer :context # :nodoc:
   attr_writer :ipv # :nodoc:
 
+  def self.rex_gethostbyname(name, resolver = @@resolver)
+    raise ::SocketError.new(
+      "Rex::Socket internal DNS resolution requires passing/setting a resolver"
+    ) unless resolver
+    # Pull both record types
+    v4 = resolver.send(name, ::Net::DNS::A).answer.first
+    v6 = resolver.send(name, ::Net::DNS::AAAA).answer.first
+    # Emulate ::Socket's error if no responses found
+    if !v4 and !v6
+      raise ::SocketError.new('getaddrinfo: Name or service not known')
+    end
+    # Build response array
+    hostbyname = [name, []]
+    if v4
+      hostbyname << ::Socket::AF_INET
+      hostbyname << self.addr_aton(v4.address.to_s)
+      hostbyname << self.addr_aton(v6.address.to_s) if v6
+    else
+      hostbyname << ::Socket::AF_INET6
+      hostbyname << self.addr_aton(v6.address.to_s)
+    end
+    return hostbyname
+  end
+
+  def self.rex_getaddrinfo(name, resolver = @@resolver)
+    raise ::SocketError.new(
+      "Rex::Socket internal DNS resolution requires passing/setting a resolver"
+    ) unless resolver
+    # Pull both record types
+    v4 = resolver.send(name, ::Net::DNS::A).answer.first
+    v6 = resolver.send(name, ::Net::DNS::AAAA).answer.first
+    # Emulate ::Socket's error if no responses found
+    if !v4 and !v6
+      raise ::SocketError.new('getaddrinfo: Name or service not known')
+    end
+    [v4, v6].compact.map do |ans|
+      ans = resolver.send(ans.name).answer.first unless ans.respond_to?(:address)
```

Comment:
Same concern here over if there is no answer, what the value of `answer` will be.

Reply:
if you compact the array and it contains `nil`s then the `do` block will never run, i think :).

Reply:
True true.

Also now that I look at it, do we need this assignment to `ans` here? It will be updated on every run of the loop so the assignment result will just be discarded?

Reply:
IIRC its a compatibility hack of some sort to make callers of the return happy. Unfortunately i'm pretty sure my memory works kind of like an ARC cache and this data slipped MFU and LRU sides of the cache :smile: - will try to divine why i would do something so inefficient.

Reply:
The inefficient code has been removed and the new code looks good so going to mark this as resolved.